rfc/egress-with-ucan.md

```diff
@@ -0,0 +1,159 @@
+
+The Gateway will offer an HTTP endpoint. Currently, the Storacha Gateway's endpoint takes the form of `https://<cid>.ipfs.w3s.link/`. The Gateway will accept a token as part of the URL. To serve the request, the Gateway will perform the following steps:
+
+1. Look up the Location Commitment for the given CID. If not found, respond with 404 Not Found.
+2. If there is no token given, serve the request through the standard rate-limiter. (This behavior is likely to change in the future.)
+3. If there is a token given (eg, `abcde12345`), build its corresponding DID (eg, `did:bearer:abcde12345`).
+4. Look up all delegations with the token DID as audience.
+5. Attempt to prove the ability to invoke `/space/content/retrieve` on the Space listed in Location Commitment, with the given CID. If more than one Location Commitment is found, attempt each in turn: a CID may be stored in multiple Spaces, and the token may be able to retrieve the content through one and not another.
+6. If no such proof chain is available, respond with 401 Unauthorized. [or 404 Not Found?]
+7. If a proof chain is found, execute the invocation on the Executor.
+8. Using the information in the receipt, fetch the content and proxy it as the response.
+
+## The Executor
+
+The [Executor](https://github.com/ucan-wg/invocation?tab=readme-ov-file#executor) of the `/space/content/retrieve` invocation will be the same service as the Gateway. [It would also be possible for Storage Nodes to run their own Executors for the central Storacha Gateway to use, but this would require more sophistication from the Storage Nodes without the benefit to them of offering their own egress service.]
+
+To execute the invocation, the Executor will perform the following steps:
+
+1. Validate the proof chain.
+2. [Look up the Location Commitment again? :/ Should it become an argument?]
+3. [Charge the Space for the egress.]
+4. [Produce a suitable receipt.]
+
+## [To Come]
+
+* Rather than serve non-token content rate-limited by default, require a delegation of `/space/content/retrieve` to some DID representing "anyone".
+* Bitswap should execute a `/space/content/retrieve` as well to respond to requests. Bitswap should be authorized by delegating to some DID representing Bitswap/Hoverboard.
+
+## Open questions
+
+* What does the `/space/content/retrieve` receipt look like?
+* Can you enumerate the contents of a space? Is that in scope here?
+* What does a Gateway URLs with a token look like?
+* What can be cached? This seems relatively cacheable, but we should be explicit in the design to make sure we're on a suitable path. This process needs to be *fast*, at least once the cache is warm.
```

Comment:
**thought**: once we've validated a token I think we can store something in [KV](https://developers.cloudflare.com/kv/) or a similarly performant caching layer, probably with an expiry, that says it's valid. we might need a mechanism to invalidate this cache, but that can probably wait until after the initial implementation?